core/controller/src/main/scala/whisk/core/controller/AuthenticatedRoute.scala

```diff
@@ -48,12 +49,43 @@ trait AuthenticatedRoute {
             }
         }
     }
-
     /** Validates credentials against database of subjects */
     protected def validateCredentials(userpass: Option[UserPass])(implicit transid: TransactionId): Future[Option[Identity]]
+
+    /** Creates client certificate auth handler */
+    protected def certificateAuth(implicit transid: TransactionId) = {
+        new CertificateAuthenticator[Identity](validateCertificate _)
+    }
+
+    /** Validates client certificate against database of subjects */
+    protected def validateCertificate(entityName: Option[String])(implicit transid: TransactionId): Future[Option[Identity]]
+
 }
 
 /** A trait for authenticated routes. */
 trait AuthenticatedRouteProvider {
     def routes(user: Identity)(implicit transid: TransactionId): Route
 }
+
+class CertificateAuthenticator[U](val userPassAuthenticator: Option[String] ⇒ Future[Option[U]])(implicit val executionContext: ExecutionContext) extends ContextAuthenticator[U] {
```

Comment:
does it make sense to just inline the apply method here into `validateCertificate` above?